### ISSUE TITLE
Question: Recommendations for benchmarking

### ISSUE THREAD

**Issue by holtjma** [varlociraptor/varlociraptor#260]
Hello,

@dlaehnemann mentioned adding varlociraptor to a benchmark we have set up in #157, and I have some questions about recommendations for setting that up.

We currently have an internal benchmark setup with but semi-recently started publishing [public results](https://github.com/HudsonAlpha/CSL_public_benchmark) in response to various discussions online.  This benchmark uses snakemake to basically set up each analysis following a `{reference}/{aligner}/{caller}` pattern with a few exceptions that we generally try to restrict as much as possible.  My understanding of varlociraptor is that it is not a de novo caller, but rather an advanced genotyper that requires a set of candidate variants (please correct me if I'm wrong about this).

Assuming the above is true, this makes me think there are two broad ways we could add the tool to our benchmark:
1. Add it as a modifier to another caller: Tthis would basically tie it to one (or more) callers (e.g. the pattern might be `{reference}/{aligner}/{caller}-varloci`). The obvious downside to this approach is that interpretation becomes more complicated, you're basically adding a fourth variable to the mix and effectively doubling the number of callers if we ran it on everything.
2. Separate it out into a genotyper analysis: In this mode, I would actually give it the full set of TP variants from the benchmark and just see how good it is at correctly genotyping the variants that we expect to exist in the data.  This is a little bit simpler to interpret and I think is more inline with how varlociraptor is intended to be used.

Or is there another approach you would recommend?

Additionally, these are all germline variants, so I'm assuming a relatively basic scenario file is what you would recommend (see below), but please correct me if you think something else would be better.
```
species:
  heterozygosity: 0.001
  ploidy:
    male:
      all: 2
      X: 1
      Y: 1
    female:
      all: 2
      X: 2
      Y: 0

samples:
  sample:
    sex: female

events:
  germline: "sample:0.5 | sample:1.0"
```

Well, I think that's probably enough to start with... thanks for your help!

**Comment by dlaehnemann** [comment]
Yes, that's exactly how varlociraptor works. You give it some set of candidate variants that you generated with some other tool, and it evaluates them with its thorough statistical model, giving you all the [flexibility of the grammar in `scenario.yaml`](https://varlociraptor.github.io/docs/calling/#generic-variant-calling). It can call ["SNVs, MNVs, indels, arbitrary replacements, inversions, duplications, and breakends in all length ranges (from small to structural)"](https://varlociraptor.github.io/landing/).

Depending on what types of variants you want to benchmark on, you need to have a respective candidate set to run varlociraptor on. If you only care about knowing the recall of confirmed variants, then your suggestion `2.` is exactly the right approach.

If you also want to know the precision, you would need to use some other caller that can generate a VCF with the respective type of variants. For example, we currently mostly use freebayes and delly, and then merge their callsets (for example see [this Snakemake workflow for varlociraptor](https://github.com/snakemake-workflows/dna-seq-varlociraptor). I that case, to get a proper recall, you would have to use those candidate-callers in a very sensitive mode, turning off as many unnecessary filters as possible. So it probably won't work to simply feed it the results of some other caller as you generated them for the benchmarking, as those would already be filtered by that caller and thus limit the recall to the recall of that caller.

In any case, you would need to filter after calling with varlociraptor. Here, [controlling the local FDR at 5%](https://varlociraptor.github.io/docs/filtering/#control-local-fdr) would be a good default, I guess.

And as for the scenario, if all you care about is a single sample and the germline variants in it, then your scenario looks right. If your sample is actually a tumor sample, and you know the tumor purity (or tumor cell content, or normal cell contamination fraction), you could actually attempt separating out germline from tumor variants. There are example scenarios for both [the case where you have further knowledge about priors](https://varlociraptor.github.io/varlociraptor-scenarios/scenarios/tumor-only-prior/) and [the case without assuming any further priors](https://varlociraptor.github.io/varlociraptor-scenarios/scenarios/tumor-only/).

Hope that already gets you somewhere, let us know if you need more infos!

**Comment by holtjma** [comment]
It sounds like option 2 is probably the most straight-forward approach overall.  A couple follow-ups after reading through this:

1. I could hand see two potentially interesting candidate sets to give varlociraptor: just the TP variant list or the TP variant plus other variants that are not in the sample (e.g. I could merge all the variants across the entire benchmark set).  The latter might be more realistic in some ways, but I'm not sure there's an obvious benefit.  Do you have any thoughts on either candidate set method?
2. In either of the above situations, is controlling the FDR is necessary or does that only apply if I'm combining a de novo caller with varlociraptor?

**Comment by dlaehnemann** [comment]
Option `2.` is definitely the easiest to implement and the quickest to run. If you want to merge in other non-TP variants to also test precision, that would be interesting to look at, as you probably also do that for other callers?

And you should definitely control for some FDR (even if it is a more permissive 0.1 or even higher), as varlociraptor calling will simply output the full list of candidate variants that you give it and will provide all the probabilities for each candidate (PROB_GERMLINE as you defined it in `germline:` and the implicit events PROB_ARTIFACT, which captures all kinds of biases and error probabilities, and PROB_ABSENT, which is basically `sample:0.0` in your case). There are no implicit filters, you really have to specify what level of false discoveries you would accept in your callset.

**Comment by holtjma** [comment]
Ah, I think I see how that plays into the setup now.  I think I have enough to go on at least for an initial crack at this, so thanks for answering all my questions!  I'll probably be in touch whenever I get an initial test to get your thoughts on the results.

**Comment by holtjma** [comment]
Hello, back with prelim results and questions.

Prelim results CSV (clipped to just have the varlociraptor results): [varloci_prelim.csv](https://github.com/varlociraptor/varlociraptor/files/8062972/varloci_prelim.csv)
Rules used for reference (and in case I royally goofed): [PipelineGenotypers.smk.txt](https://github.com/varlociraptor/varlociraptor/files/8063009/PipelineGenotypers.smk.txt)

So looking at those results, there are more false negatives than I was expecting and it doesn't seem to be contained to an individual sample (I checked that the performance is roughly consistent across the various samples).  Is this where you expected performance to be?  If not, how would you recommend going about debugging?

**Comment by johanneskoester** [comment]
Hi @holtjma, thanks a lot for having a look at this! We are about to make our own GIAB benchmarking public as well. Your general impression about the recall is analogous to what we see, but it is important to note that Varlociraptor has some (to the best of my knowledge) unique properties regarding the precision in particular in complex regions. Basically, it boils down to the fact that the model manages to accurately quantify the uncertainty equally well for both high coverage unique as well as low coverage and highly ambiguous loci. This comes at the cost to being sometimes a bit more conservative than other callers. One can see this best when stratifying calls by coverage and also looking at precision simultaneously, but it will be easiest to discuss that in person.

@dlaehnemann will contact you about a personal meeting.

**Comment by johanneskoester** [comment]
@holtjma we weren't able to reach you, neither via twitter nor via mail. Can you drop me a mail instead? You find my mail contact here: https://koesterlab.github.io

**Comment by holtjma** [comment]
Strange, not sure why those aren't working.  I just sent an email, so hopefully that will work to start a comms channel.

**Comment by jdidion** [comment]
Hi @johanneskoester, just curious if there is a timeline for your GIAB results. I'm interested to see how you've performed the analysis and how well varlociraptor does vs other callers. Thanks!
